torch/nn/modules/module.py

```diff
@@ -1309,19 +1372,43 @@ def state_dict(self, destination=None, prefix='', keep_vars=False):
             ['bias', 'weight']
 
         """
-        if destination is None:
+
+        # Fixes https://github.com/pytorch/pytorch/issues/72778
+        # If calling with overload 1, everything should be fine whatsoever
+        # If calling with overload 2, there will possibly be mismatch
+        # Case 1: use default value or kwargs only
+        #   Nothing wrong
+        #
+        # Case 2: only prefix is positional
+        #       mod.state_dict(pref) or mod.state_dict(pref, keep_vars=keep)
+        #   Actual values are:
+        # destination=pref, prefix='', keep_vars={False|keep}
+        #
+        # Case 3: both positional
+        #       mod.state_dict(pref, keep)
+        #   Actual values are:
+        # destination=pref, prefix=keep, keep_vars=False
+
+        if isinstance(destination, torch._six.string_classes):
```

Comment:
Not sure what you're trying to do here.
The two arguments will have to be passed as kwargs in the final version of the function.
You can mark this with kwarg-only arguments: `def state_dict(self, *, prefix: str = ..., keep_vars: bool = ...)`.

Reply:
I thought it was the safest way to fix it by keeping both the original signatures valid.
Moving directly to kw-only can be an easier fix and I agree it will be the final version. However, it may break BC, e.g. some existing code could be using `mod.state_dict(some_custom_dict, 'foo.', True)`.
I need further confirmation on what signature should be exposed to users in this fix.

Reply:
Existing codes using `(prefix, keep_vars)` have no BC issues, because it would cause a crash if not using kwargs.

Existing code using `(destination, prefix, keep_vars)` can have positional args, so forcing kw-args may cause problems. However, this is the use case that we want to deprecate.

Reply:
A possible solution:
```python
def state_dict(destination, _prefix, _keep_vars, *, prefix, keep_vars):
    if prefix is provided: use prefix
    elif _prefix is provided: use prefix=_prefix
    else: use prefix=default
    # same for keep_vars
    if any destination, _prefix, _keep_vars: warning
```

Reply:
> I thought it was the safest way to fix it by keeping both the original signatures valid.

I think that unfortunately, the current overload annotations are wrong. And in particular, the second one is not valid. So that is fine to not support that (since it never worked).

For the fix, what about only warning today if "destination" is passed? Since the only way not to pass in destination is to pass the two others as kwargs. Once we delete destination, we can make the two others kwargs only.

So I think what we want to do here is what you have for the current signature
```python
def state_dict(*, prefix, keep_vars): ...
def state_dict(destination, prefix, keep_vars): ...  # This one is deprecated!
```

And the implementation can simply be
```python
def state_dict(self, *args, prefix='', keep_vars=False):
    if len(args)>0:
        # warn here that positional is deprecated
        destination = args[0]
        if len(args)>1:
            # check that prefix == ''
            prefix = args[1]
        # same for keeps var
        
```

What do you think?